Providers/Profile.swift

```diff
@@ -576,28 +592,48 @@ public class BrowserProfile: Profile {
         }
 
         /**
-         * Locking is managed by withSyncInputs. Make sure you take and release these
+         * Locking is managed by syncSeveral. Make sure you take and release these
          * whenever you do anything Sync-ey.
          */
-        var syncLock = OSSpinLock() {
-            didSet {
-                let notification = syncLock == 0 ? NotificationProfileDidFinishSyncing : NotificationProfileDidStartSyncing
-                NSNotificationCenter.defaultCenter().postNotification(NSNotification(name: notification, object: nil))
-            }
-        }
+        private let syncLock = NSRecursiveLock()
 
-        // According to the OSAtomic header documentation, the convention for an unlocked lock is a zero value
-        // and a locked lock is a non-zero value
         var isSyncing: Bool {
-            return syncLock != 0
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            return !(syncTerminal?.isFilled ?? true)
         }
 
+        private let syncQueue = dispatch_queue_create("com.mozilla.firefox.sync", DISPATCH_QUEUE_SERIAL)
+
         private func beginSyncing() -> Bool {
-            return OSSpinLockTry(&syncLock)
+            let ready = !isSyncing
+            if ready {
+                notifySyncing(NotificationProfileDidStartSyncing)
+            }
+            return ready
+        }
+
+        private func endSyncingMaybe(statuses: [EngineStatus] = []) {
+            // Sync `statuses` may have come from currentSync, or one before it.
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            // If currentSync (the last sync job called) hasn't been filled, then we're still syncing.
+            let stillSycing = !(currentSync?.isFilled ?? true)
+            if !stillSycing {
+                endSyncing(statuses)
+            }
         }
 
-        private func endSyncing() {
-            OSSpinLockUnlock(&syncLock)
+        private func endSyncing(statuses: [EngineStatus]) {
```

Comment:
Document all of these private functions: they must only be called while the sync lock is held.